glean-core/src/ping/mod.rs

```diff
@@ -67,12 +70,80 @@ impl PingMaker {
         current_seq as usize
     }
 
+    fn get_start_end_times(&self, glean: &Glean, storage_name: &str) -> (Metric, Metric) {
+        let time_unit = TimeUnit::Minute;
+
+        let start_time = DatetimeMetric::new(
+            CommonMetricData {
+                name: format!("{}#start", storage_name),
+                category: "".into(),
+                send_in_pings: vec![INTERNAL_STORAGE.into()],
+                lifetime: Lifetime::User,
+                ..Default::default()
+            },
+            time_unit,
+        );
+
+        let end_time = DatetimeMetric::new(
+            CommonMetricData {
+                name: format!("{}#end", storage_name),
+                category: "".into(),
+                send_in_pings: vec![INTERNAL_STORAGE.into()],
+                lifetime: Lifetime::User,
+                ..Default::default()
+            },
+            time_unit,
+        );
+
+        let start_time_data = match StorageManager.snapshot_metric(
+            glean.storage(),
+            INTERNAL_STORAGE,
+            &start_time.meta().identifier(),
+        ) {
+            Some(Metric::Datetime(t, tu)) => {
+                log::info!("datetime start found: {:?}", t);
+                Metric::Datetime(t, tu)
+            }
+            _ => {
+                let now: DateTime<Local> = Local::now();
+                let naive = now.naive_utc();
+                let fixed_tz = Local.offset_from_utc_datetime(&naive);
+                let now_fixed: DateTime<FixedOffset> = fixed_tz.from_utc_datetime(&naive);
+                log::info!("datetime start missing, creating now: {:?}", now_fixed);
+                Metric::Datetime(now_fixed, time_unit)
+            }
+        };
+
+        let now: DateTime<Local> = Local::now();
+        let naive = now.naive_utc();
+        let fixed_tz = Local.offset_from_utc_datetime(&naive);
+        let now_fixed: DateTime<FixedOffset> = fixed_tz.from_utc_datetime(&naive);
+
+        log::info!("datetime end setting to now: {:?}", now_fixed);
+        end_time.set(glean, now_fixed);
+
+        let end_time_data = match StorageManager.snapshot_metric(
+            glean.storage(),
+            INTERNAL_STORAGE,
+            &end_time.meta().identifier(),
+        ) {
+            Some(Metric::Datetime(t, tu)) => {
+                start_time.set(glean, t);
+                Metric::Datetime(t, tu)
+            }
+            _ => unreachable!("End time just set, but now missing"),
```

Comment:
Is this possible in practice? Does the `unreachable!` string get printed to the logs in case of problems? I'd like it to show on CI logs, if this ever happens